ocaml/xapi/xapi_vm.ml

```diff
@@ -238,6 +238,29 @@ let assert_memory_constraints ~__context ~vm platformdata =
    same VM twice during memory calculations to determine whether a given VM can start on a particular host..
 *)
 
+let update_platform_secureboot ~__context ~self platform =
+  if List.assoc "secureboot" platform = "auto" then
+    let t = ("secureboot", string_of_bool (Db.Pool.get_uefi_certificates ~__context ~self <> "")) in
+    List.append [t] (List.remove_assoc "secureboot" platform)
+  else
+    platform
+
+let save_uefi_certificates_to_dir ~__context ~self =
+  let dir = ref [||] in
+  try
+    dir := Sys.readdir !Xapi_globs.varstore_dir;
+  with Sys_error _ ->
+    begin
+      debug "varstored directory does not exist";
+      Xapi_stdext_unix.Unixext.mkdir_rec "/usr/share/varstored" 0o755
+    end;
+
+  if not (Array.exists (fun e -> e = "PK.auth") !dir) then
+    let file = Unix.openfile "xapi_uefi_certificates.tar" [Unix.O_RDWR; Unix.O_CREAT] 0o755 in
+    let contents = (Xapi_stdext_base64.Base64.decode (Db.Pool.get_uefi_certificates ~__context ~self)) in
+    let hdr = Tar_unix.Header.make "xapi_uefi_certificates.tar" (Int64.of_int (String.length contents)) in
+    Tar_unix.write_block hdr (fun file -> Unixext.really_write_string file contents) file
```

Comment:
You need to extract the tarfile, but this one creates another tar file.
Instead you should use Tar_unix.extract or extract_gen, see https://github.com/mirage/ocaml-tar/blob/4a6cad9a5e947e5b7c5953c3e4f6a9535a0f593a/unix/tar_unix.ml#L99-L105